drivers/adc/adc_nrfx_saadc.c

```diff
@@ -0,0 +1,374 @@
+
+	if (!repeat) {
+		nrf_saadc_buffer_pointer_set(
+			nrf_saadc_buffer_pointer_get() +
+			nrf_saadc_amount_get());
+	}
+}
+
+static int set_resolution(const struct adc_sequence *sequence)
+{
+	nrf_saadc_resolution_t nrf_resolution;
+
+	switch (sequence->resolution) {
+	case  8:
+		nrf_resolution = NRF_SAADC_RESOLUTION_8BIT;
+		break;
+	case 10:
+		nrf_resolution = NRF_SAADC_RESOLUTION_10BIT;
+		break;
+	case 12:
+		nrf_resolution = NRF_SAADC_RESOLUTION_12BIT;
+		break;
+	case 14:
+		nrf_resolution = NRF_SAADC_RESOLUTION_14BIT;
+		break;
+	default:
+		SYS_LOG_ERR("ADC resolution value %d is not valid",
+			    sequence->resolution);
+		return -EINVAL;
+	}
+
+	nrf_saadc_resolution_set(nrf_resolution);
+	return 0;
+}
+
+static int set_oversampling(const struct adc_sequence *sequence,
+			    u8_t active_channels)
+{
+	nrf_saadc_oversample_t nrf_oversampling;
+
+	if ((active_channels > 1) && (sequence->oversampling > 0)) {
+		SYS_LOG_ERR(
+			"Oversampling is supported for single channel only");
+		return -EINVAL;
+	}
+
+	switch (sequence->oversampling) {
+	case 0:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_DISABLED;
+		break;
+	case 1:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_2X;
+		break;
+	case 2:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_4X;
+		break;
+	case 3:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_8X;
+		break;
+	case 4:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_16X;
+		break;
+	case 5:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_32X;
+		break;
+	case 6:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_64X;
+		break;
+	case 7:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_128X;
+		break;
+	case 8:
+		nrf_oversampling = NRF_SAADC_OVERSAMPLE_256X;
+		break;
+	default:
+		SYS_LOG_ERR("Oversampling value %d is not valid",
+			    sequence->oversampling);
+		return -EINVAL;
+	}
+
+	nrf_saadc_oversample_set(nrf_oversampling);
+	return 0;
+}
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(nrf_saadc_value_t);
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		SYS_LOG_ERR("Provided buffer is too small (%u/%u)",
+			    sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	int error = 0;
+	u32_t channels = sequence->channels;
+	u8_t active_channels = 0;
+	u8_t channel_id;
+
+	/* Signal an error if channel selection is invalid (no channels or
+	 * a non-existing one is selected).
+	 */
+	if (!channels || (channels & (~0UL << NRF_SAADC_CHANNEL_COUNT))) {
+		SYS_LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	active_channels = 0;
+
+	/* Enable only the channels selected for the pointed sequence.
+	 * Disable all the rest.
+	 */
+	channel_id = 0;
+	do {
+		if (channels & 1) {
```

Comment:
seems like the 1 here could be replaced w/define or some support.  Is this an enable bit?